File: app/src/main/java/com/gnoemes/shikimoriapp/entity/anime/data/AnimeDetailsResponse.java
package com.gnoemes.shikimoriapp.entity.anime.data;

import android.support.annotation.Nullable;

import com.gnoemes.shikimoriapp.entity.rates.data.UserRateResponse;
import com.google.gson.annotations.SerializedName;

import org.joda.time.DateTime;

import java.util.List;

/**
 * Anime full info response data model
 */
public class AnimeDetailsResponse {

    @SerializedName("id")
    private long id;

    @SerializedName("topic_id")
    private long topicId;

    @SerializedName("name")
    private String name;

    @Nullable
    @SerializedName("russian")
    private String russianName;

    @SerializedName("image")
    private AnimeImageResponse image;

    @SerializedName("url")
    private String url;

    @SerializedName("kind")
    private String type;

    @Nullable
    @SerializedName("status")
    private String status;

    @SerializedName("episodes")
    private int episodes;

    @SerializedName("episodes_aired")
    private int episodesAired;

    @Nullable
    @SerializedName("aired_on")
    private DateTime airedDate;

    @Nullable
    @SerializedName("released_on")
    private DateTime releasedDate;

    @Nullable
    @SerializedName("english")
    private List<String> englishNames;

    @Nullable
    @SerializedName("japanese")
    private List<String> japaneseNames;

    @SerializedName("duration")
    private int duration;

    @SerializedName("score")
    private double score;

    @Nullable
    @SerializedName("description")
    private String description;

    @SerializedName("genres")
    private List<GenreResponse> genres;

    @SerializedName("user_rate")
    private UserRateResponse rateResponse;

    public long getId() {
        return id;
    }

    public long getTopicId() {
        return topicId;
    }

    public String getName() {
        return name;
    }

    @Nullable
    public String getRussianName() {
        return russianName;
    }

    public AnimeImageResponse getImage() {
        return image;
    }

    public String getUrl() {
        return url;
    }

    public String getType() {
        return type;
    }

    @Nullable
    public String getStatus() {
        return status;
    }

    public int getEpisodes() {
        return episodes;
    }

    public int getEpisodesAired() {
        return episodesAired;
    }

    @Nullable
    public DateTime getAiredDate() {
        return airedDate;
    }

    @Nullable
    public DateTime getReleasedDate() {
        return releasedDate;
    }

    @Nullable
    public List<String> getEnglishNames() {
        return englishNames;
    }

    @Nullable
    public List<String> getJapaneseNames() {
        return japaneseNames;
    }

    public int getDuration() {
        return duration;
    }

    public double getScore() {
        return score;
    }

    @Nullable
    public String getDescription() {
        return description;
    }

    public List<GenreResponse> getGenres() {
        return genres;
    }

    public UserRateResponse getRateResponse() {
        return rateResponse;
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/entity/anime/data/AnimeDetailsResponse.kt
package com.gnoemes.shikimoriapp.entity.anime.data

import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeStatus
import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeType
import com.gnoemes.shikimoriapp.entity.common.data.ImageResponse
import com.gnoemes.shikimoriapp.entity.rates.data.UserRateResponse
import com.gnoemes.shikimoriapp.entity.video.data.VideoResponse
import com.google.gson.annotations.SerializedName
import org.joda.time.DateTime

data class AnimeDetailsResponse(
        @field:SerializedName("id") val id: Long,
        @field:SerializedName("name") val name: String,
        @field:SerializedName("russian") val nameRu: String?,
        @field:SerializedName("image") val image: ImageResponse,
        @field:SerializedName("url") val url: String,
        @field:SerializedName("kind") val type: AnimeType,
        @field:SerializedName("status") private val _status: AnimeStatus?,
        @field:SerializedName("episodes") val episodes: Int,
        @field:SerializedName("episodes_aired") val episodesAired: Int,
        @field:SerializedName("aired_on") val dateAired: DateTime?,
        @field:SerializedName("released_on") val dateReleased: DateTime?,
        @field:SerializedName("rating") val rating: String,
        @field:SerializedName("english") val namesEnglish: List<String?>?,
        @field:SerializedName("japanese") val namesJapanese: List<String?>?,
        @field:SerializedName("duration") val duration: Int,
        @field:SerializedName("score") val score: Double,
        @field:SerializedName("description") val description: String?,
        @field:SerializedName("description_html") val descriptionHtml: String,
        @field:SerializedName("favoured") val isFavorite: Boolean,
        @field:SerializedName("topic_id") val topicId: Long,
        @field:SerializedName("genres") val genres: List<GenreResponse>,
        @field:SerializedName("videos") val videos: List<VideoResponse>,
        @field:SerializedName("user_rate") val userRate: UserRateResponse?
) {

    val status
        get() = _status ?: AnimeStatus.NONE

}